Lib/unittest/mock.py

```diff
@@ -925,10 +926,10 @@ def assert_any_call(self, /, *args, **kwargs):
         The assert passes if the mock has *ever* been called, unlike
         `assert_called_with` and `assert_called_once_with` that only pass if
         the call is the most recent one."""
-        expected = self._call_matcher((args, kwargs))
+        expected = self._call_matcher(_Call((args, kwargs), two=True))
+        cause = expected if isinstance(expected, Exception) else None
         actual = [self._call_matcher(c) for c in self.call_args_list]
-        if expected not in actual:
-            cause = expected if isinstance(expected, Exception) else None
+        if cause or expected not in _AnyComparer(actual):
```

Comment:
I think as per my commit we can remove `_AnyComparer`. It takes the expected value which could contain ANY and make sure it stays on the left while comparing with actual. The calllist `__contains__` already does the same check here with https://github.com/python/cpython/blob/455122a0094c8cfdf7e062eccc5e5b5885c75e8b/Lib/unittest/mock.py#L340. The list `__contains__` does an equal per item and hence it does the same check at `_Call.__eq__` https://github.com/python/cpython/blob/455122a0094c8cfdf7e062eccc5e5b5885c75e8b/Lib/unittest/mock.py#L2394

There are also no test failures and I see the behavior around comparison to be the same. We also can skip the check for cause as I go with the approach. @ElizabethU  Thoughts?

Reply:
@tirkarthi I'm all for removing `_AnyComparer` as I think it's pretty clunky, but I'm seeing test failures with [tirkarthi@bc09989]. First I cherry-picked your commit onto my branch and saw test failures, then I tried on the branch on your fork, thinking that you might have had some other changes or even that someone else might have made some changes that you'd have if you'd synced your fork since I'd opened this PR. I'm still getting the same 4 failures either way:

`unittest.test.testmock.testasync.AsyncMockAssert.test_awaits_asserts_with_any`
`unittest.test.testmock.testasync.AsyncMockAssert.test_awaits_asserts_with_spec_and_any`
`unittest.test.testmock.testhelpers.AnyTest.test_any_and_spec_set`
`unittest.test.testmock.testhelpers.AnyTest.test_any_no_spec`

Are those four tests working for you? If so, I'm pretty confused since these failures really make sense to me. Firstly, in `assert_any_call` above, whether `actual` is just a regular list or a `_CallList`, it goes into the default list class's `__contains__` either way. And when it does, `call(<ANY>, 1)` is the `value`, and it gets a little confusing, but the way item by item comparison is done, when it enters `_Call.__eq__` https://github.com/python/cpython/blob/455122a0094c8cfdf7e062eccc5e5b5885c75e8b/Lib/unittest/mock.py#L2394
`self` is `call(<ANY>, 1)` and `other` is an item from the list, meaning that when it hits the reversed comparison at the end of the method
https://github.com/python/cpython/blob/455122a0094c8cfdf7e062eccc5e5b5885c75e8b/Lib/unittest/mock.py#L2444
`ANY` is on the right, and it doesn't work.

At least that's what I'm seeing on my machine. I'd be really excited if you've found a way around `_AnyComparer`, but it seems like we still need it.

Reply:
Confirmed that this change in the order of comparison to be crucial to my patch over change in the arguments. That means we can either 

1. Have ANYCompare and apply this fix on 3.9 and 3.8 . I am not sure of backporting this to 3.7 at this point since it's in 3.7.5 but I will leave it upto the core dev.
2. Depend on the below order in list comparison and remove ANYCompare to use just contains and keep the fix only to 3.9 since the listobject patch is not on 3.8 and below.

Thoughts?

```
diff --git a/Objects/listobject.c b/Objects/listobject.c
index d012ab933a..cea9b24a3b 100644
--- a/Objects/listobject.c
+++ b/Objects/listobject.c
@@ -449,8 +449,7 @@ list_contains(PyListObject *a, PyObject *el)
     int cmp;

     for (i = 0, cmp = 0 ; cmp == 0 && i < Py_SIZE(a); ++i)
-        cmp = PyObject_RichCompareBool(el, PyList_GET_ITEM(a, i),
-                                           Py_EQ);
+        cmp = PyObject_RichCompareBool(PyList_GET_ITEM(a, i), el, Py_EQ);
     return cmp;
 }
```

Reply:
Thinking a bit further there is also mock backport which would need to run on 3.6+ so this would cause problems. Now I am more leaning towards to reverting my change to keep ANYCompare though the interpreter is fixed in 3.9+. At this point I think I depended on the patch in 3.9 during debugging and made some incorrect assumptions with respect to ordering, sorry about that. I would be okay with this landing on 3.8+ with ANYcompare and the backport can work on 3.6+ with the PR.

@ElizabethU Feel free to revert my patch.

Reply:
I am able to get it to work by rebasing 🎉 🎊 🎺 Thanks, I should have tried that earlier, but I thought since I was still having the problem on your branch and you had that passing, that it couldn't be a rebase. I should have tried anyway.

My input on the patch for 3.9 is that it's great and a big improvement over what I had alone and I would love to see it get reviewed whenever there's time. As for backporting, maybe we'd still need the `_AnyComparer` class for 3.8 then? Or maybe this just doesn't get backported? Not sure what's the norm, but I'm happy to learn and go along with it.

Thank you so, so much for sticking with me through this. I've learned so much, and your expertise has been a big part of that. I hope I can contribute again and take up a lot less of your time now that I know the routine and some of the codebase.

Reply:
@tirkarthi Ah, I didn't see your last comments. So reverting is the right thing to do? I do like the patch, and was hoping we could keep it for 3.9, but I'm unfamiliar with the mock backport situation, and why it means we couldn't use the patch in just 3.9?

I'm reverting the patch per your instructions.

Reply:
FWIW, anything that means things work on 3.6+ would be very much appreciated by me :-)